### PR TITLE
Go 1.21 Parallelism and Test Fixes

### PR DESCRIPTION
**Pull request by Otterverse**
tl;dr This attempts to fix some issues around using go 1.21 and testing.

Problem 1: Go 1.21 now imports and init()s packages in a well-defined order. This breaks some stuff in vision that was using init() to set global vars during testing. This attempts to fix those issues as they are effectively "bugs."

Problem 2: A secondary problem is that go test with "-coverprofile" enabled runs *very* slow when combined with "-race" and that problem magnifies the more concurrent processes there are. To fix this, the test.sh and makefile are updated to run seperate -race and then -coverprofile test runs.

These sum to the fact that, due to *bad* behavior happening (e.g. one test/file was setting a global, and it was being applied to ALL tests in the package) we were getting away with the coverage tests not slowing as much. When go1.21 changed things, it exposed this problem, as the segmenter tests were now running truly parallel... which made coverage much slower, even though regular go tests (with OR without -race) are much faster.

The overall test times (with the now-split test runs) are just about the same overall as the previous (wrongly behaving) cover+race test. The good news is, these can/should potentially be split in CI, so they can run in parallel, and can likely reduce the overall CI testing time (in terms of human wait.)

```
This PR go119
#go clean -testcache && time make cover
real	6m25.573s
user	52m59.433s
sys	2m27.864s
```
```
This PR go121
#go clean -testcache && time make cover
real	7m12.095s
user	62m22.795s
sys	2m17.161s
```
```
Tihis PR go121
#go clean -testcache && time make test-go
real	5m16.975s
user	46m52.478s
sys	1m24.302s
```
```
This PR go121 (hacked to not run test-go first as part of "cover")
#go clean -testcache && time make cover
real	2m44.033s
user	18m49.432s
sys	0m55.633s
```
```
Main go119
#go clean -testcache && time make cover (FAILED repeatedly on my machine, due to a bug fix included here)
real	5m51.762s
user	43m35.917s
sys	1m34.114s
```
```
Main go121
#go clean -testcache && time make cover
real	9m51.232s
user	47m31.321s
sys	1m49.398s
```

## test.yml changes
- I parallelized `make cover-only` and `make test-go` per James' speedup suggestion
- see screenshots below for timing. tldr, an arbitrary run of main branch took 20:33 in its longest step, this takes about 15
- sample run here https://github.com/viamrobotics/rdk/actions/runs/7205015526

image

mainyml

### CODE REVIEW

File: vision/segmentation/plane_segmentation_test.go [hunk 1]
```diff
@@ -51,6 +53,12 @@ func TestPlaneConfig(t *testing.T) {
 }
 
 func TestSegmentPlaneWRTGround(t *testing.T) {
+	// Setting a global, so cannot use t.Parallel()
+	sortPositions = true
+	defer func() {
+		sortPositions = false
+	}()
```
Comment (Otterverse) [inline vision/segmentation/plane_segmentation_test.go:60]: I did this in all the tests of this file, since it was originally a global, but I'm not sure it's all needed.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -89,9 +112,71 @@ jobs:
...
+    - name: Upload code coverage
+      uses: actions/upload-artifact@v3
+      with:
+        name: pr-code-coverage
+        path: |
+          pr.env
+          code-coverage-results.md
+        retention-days: 1
+
+  comments:
```
Comment (Otterverse) [inline .github/workflows/test.yml:174]: This doesn't need to be it's own job. Just handle it in the "all tests passing" step for efficiency, since these variables/output won't be used unless that job passes anyway.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -192,7 +268,7 @@ jobs:
 
   test_passing:
     name: All Tests Passing
-    needs: [test_pi, build_and_test, test_e2e, test32]
+    needs: [test_pi, comments, test_e2e, test32]
     runs-on: [ubuntu-latest]
     if: always()
     steps:
```
[the remark this answers is not in the document]
Comment (Otterverse) [reply .github/workflows/test.yml:274]: I see you fixed this on the main branch, so just remember to pull that in to this PR.

**Review by Otterverse — COMMENTED** [review]
So while we're trying to re-arrange things to be more parallel, I've got a suggestion to rearrange/group tests.


build-go/lint-go takes about 2 minutes, and as the spinup time for a new runner is about 45s to a minute, it's a small penalty to leave it in the same Job as the test run (and it saves build time slightly.) But we don't need to test all the frontend stuff on every architecture either.

Splitting the web stuff out makes it clearer where failures are, but again, as build/lint here is fast (30-60 seconds) and test-web does nothing, I'd suggest amending test-e2e onto the same Job to save spin-up time, since e2e has to build-web anyway.

Thus, I suggest the following split-up of tests. All of these would be checked by the "all-tests-passing" job/block point.

```
# Backend/Go
# amd64/arm64/armv7 (matrix)
# armv7 might continue to skip build/lint as it does now
make build-go lint-go
AND
make test-go
```
```
# Frontend/JS
# Note: test-web does nothing right now, as we have no unit tests.
# amd64 only
make build-web lint-web
AND
make test-web test-e2e
```
```
# arm64 PI only
make test-pi
```

File: Makefile [hunk 4]
```diff
@@ -73,6 +73,9 @@ check-web:
 cover: test-go
 	PATH=$(PATH_WITH_TOOLS) ./etc/test.sh cover
 
+cover-only: tool-install
```
[the remark this answers is not in the document]
Comment (abe-winter) [reply Makefile:73]: done in b8b905bb

File: .github/workflows/test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (abe-winter) [reply .github/workflows/test.yml:227]: oof I switched to a branch on origin to test the workflow and forgot the PR was still pointing at the old one. both are now synced at 398b3cfa

File: .github/workflows/test.yml (hunk 2 above)
Comment (abe-winter) [reply .github/workflows/test.yml]: done in 103f8c7c

**Comment by abe-winter** [comment]
your edited splits took us from ~20 minutes to ~15 minutes

(I attached lint to coverage, instead of race, because coverage is slightly faster)

e2e is failing now because some apt installs are 404ing; I added an `apt-get update` in the e2e step to fix that, but the run attached to the PR isn't using the updated yaml. successful run here https://github.com/viamrobotics/rdk/actions/runs/7205015526

**Comment by Otterverse** [comment]
Okay, ready for final review, @abe-winter Pushed to both this and the upstream branch. Test run is here: https://github.com/viamrobotics/rdk/actions/runs/7213547953

**Review by abe-winter — APPROVED** [review]
